src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs

```diff
@@ -39,6 +39,38 @@ public CertificateCallbackMapper(Func<HttpRequestMessage, X509Certificate2, X509
         }
 
         public static async ValueTask<(Socket, Stream)> ConnectAsync(string host, int port, CancellationToken cancellationToken)
+        {
+            CancellationTokenSource cancelIPv6 = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            Task<(Socket, Stream)> tryConnectAsyncIPv6 = ConnectAsyncInternal(host, port, AddressFamily.InterNetworkV6, cancelIPv6.Token);
+
+            if (await Task.WhenAny(tryConnectAsyncIPv6, Task.Delay(200)).ConfigureAwait(false) == tryConnectAsyncIPv6 && tryConnectAsyncIPv6.IsCompletedSuccessfully)
```

Comment:
I skimmed the happy eyeballs RFC, and it seems to suggest a different approach than this is taking.  It states:
```
   Next, the client SHOULD modify the ordered list to interleave address
   families.  Whichever address family is first in the list should be
   followed by an address of the other address family; that is, if the
   first address in the sorted list is IPv6, then the first IPv4 address
   should be moved up in the list to be second in the list.  An
   implementation MAY want to favor one address family more by allowing
   multiple addresses of that family to be attempted before trying the
   other family.  The number of contiguous addresses of the first
   address family will be referred to as the "First Address Family
   Count" and can be a configurable value.  This is performed to avoid
   waiting through a long list of addresses from a given address family
   if connectivity over that address family is impaired.
```
I don't know if your approach is better or worse, but it seems different.

Reply:
I'll give the spec a more thorough read to make sure I'm taking the right approach.

Reply:
You're right, this is different. The implementation here has two separate queues of addresses with one address from each queue attempting to connect simultaneously. It only has one 200 ms delay, to start attempting IPv6 before starting IPv4. The spec actually suggests starting a new connection every 200 ms, alternating between the queues. This greatly increases the upper bound on the number of simultaneous connections we will attempt, but allows us to be more robust on the case where many of the addresses returned by the DNS are not functional.

I'll make sure my implementation matches up to the spec.